### PR TITLE
Robot to Node April Tag angle calculation added

### PR DESCRIPTION
**Pull request by MG-05**

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -54,17 +53,26 @@ public class Constants {
 	public static final double HUB_Y_COORDINATE = 0;
 
 	public static final double QUARTER_PI = Math.PI / 4;
+	public static final double METERS_TO_INCHES_CONSTANT = 39.3701;
```
Comment (AG6GR) [inline src/main/java/frc/robot/Constants.java:56]: Use Units.metersToInches

File: src/main/java/frc/robot/TeleopInput.java [hunk 2]
```diff
@@ -27,6 +26,9 @@ public class TeleopInput {
 	private static final int AIM_LOW_BUTTON = 11;
 	private static final int SUBSTATION_PICKUP_BUTTON = 8;
 	private static final int HOMING_BUTTON = 10;
+	private static final int CV_ALLIGN_BUTTON_LEFT_NODE = 4;
```
Comment (AG6GR) [inline src/main/java/frc/robot/TeleopInput.java:29]: Spelling. Also please add to the teleop button map spreadsheet.

File: src/main/java/frc/robot/systems/DriveFSMSystem.java [hunk 3]
```diff
@@ -180,12 +201,20 @@ public void update(TeleopInput input) {
 				handleTeleOpBalanceState(input);
 				break;
 
+			case TELE_STATE_CV_ALLIGN:
+				xToATag = pcw.getEstimatedGlobalPose().get().estimatedPose.getX()
```
Comment (AG6GR) [inline src/main/java/frc/robot/systems/DriveFSMSystem.java:205]: Move these into the handleCVAllignState function.

File: src/main/java/frc/robot/systems/DriveFSMSystem.java [hunk 4]
```diff
@@ -424,6 +463,27 @@ private void handleTeleOp2MotorState(TeleopInput input) {
 			leftPower = power.getLeftPower();
 			rightPower = power.getRightPower();
 
+			if (!pcw.getEstimatedGlobalPose().isEmpty()) {
+				// left is negative right is positive
+				angleToTurnToFaceTag = -1 * (Constants.ONE_REVOLUTION_DEGREES - Math.toDegrees(
```
Comment (AG6GR) [inline src/main/java/frc/robot/systems/DriveFSMSystem.java:468]: These couple lines would be a lot more readable if you stored `pcw.getEstimatedGlobalPose().get().estimatedPose` to a function local variable.

File: src/main/java/frc/robot/systems/DriveFSMSystem.java [hunk 5]
```diff
@@ -424,6 +463,27 @@ private void handleTeleOp2MotorState(TeleopInput input) {
...
+					pcw.getEstimatedGlobalPose().get().estimatedPose.getRotation().getAngle()));
+
+				if (pcw.getEstimatedGlobalPose().get().estimatedPose.getY() < 0) {
```
Comment (AG6GR) [inline src/main/java/frc/robot/systems/DriveFSMSystem.java:471]: Is this just doing absolute value? I think if you simplify the math you don't need the if else.

**Review by AG6GR — CHANGES_REQUESTED** [review]
Just FYI, "align" has one L.

File: src/main/java/frc/robot/systems/DriveFSMSystem.java [hunk 6]
```diff
@@ -161,15 +172,26 @@ public void resetTeleop() {
...
 		dpe.updatePose(gyro.getAngle(), leftMotor.getEncoder().getPosition(),
 			rightMotor.getEncoder().getPosition());
-		SmartDashboard.putNumber("X", dpe.getCurPose().getX());
-		SmartDashboard.putNumber("Y", dpe.getCurPose().getY());
-		SmartDashboard.putNumber("Rotation", dpe.getCurPose().getRotation().getDegrees());
-		// gyroAngleForOdo = gyro.getAngle();
+
+		if (!pcw.getEstimatedGlobalPose().isEmpty()) {
+			SmartDashboard.putNumber("X",
+				pcw.getEstimatedGlobalPose().get().estimatedPose.getX()
```
Comment (AG6GR) [inline src/main/java/frc/robot/systems/DriveFSMSystem.java:178]: I meant we shouldn't need METERS_TO_INCHES_CONSTANT, just call the conversion function instead.

**Review by AG6GR — APPROVED** [review]

**Review by mirapanigrahy — APPROVED** [review]

**Pull request merged**